### PR TITLE
Smartcontract types definition in interops

### LINKED ISSUE #1372: Smartcontract types definition in interops
Contract's manifest describes methods using smartcontract parameter types with all its "signatures", "hash160" and alike. But we autogenerate them based on Go types and at the moment we just don't have types corresponding to all smartcontract parameter types. We should add some definitions like 
```
type Signature []byte
type Hash160 []byte
```
to `engine` or `runtime` package and adjust manifest generator to emit proper ABI types when contract uses them.

### PR DESCRIPTION
**Pull request by fyrchik**
Close #1372 .

### CODE REVIEW

File: pkg/interop/types.go [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package interop
```
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/interop/types.go:1]: We're also missing an `InteropInterface` here. Most probably it's not explicitly needed, but I'm not sure we're generating it properly for things like `storage.Context` or `iterator.Iterator`.

[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/interop/types.go:1]: I am not sure, that `InteropInterface` will ever be presented in a manifest: it is used internally and can't be marshalled/unmarshalled, so it is impossible to call such method.

[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/interop/types.go:1]: It doesn't need to be marshalled/unmarshalled. One contract can take a storage context or iterator/enumerator from the other contract as a parameter, why not?

[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/interop/types.go:1]: Fixed.

[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/interop/types.go:1]: Fixed.

File: pkg/interop/blockchain/blockchain.go [hunk 2]
```diff
@@ -3,21 +3,24 @@ Package blockchain provides functions to access various blockchain data.
...
+	"github.com/nspcc-dev/neo-go/pkg/interop/contract"
+)
 
 // Transaction represents a NEO transaction. It's similar to Transaction class
 // in Neo .net framework.
 type Transaction struct {
 	// Hash represents the hash (256 bit BE value in a 32 byte slice) of the
 	// given transaction (which also is its ID).
-	Hash []byte
+	Hash interop.Hash256
```
Comment (AnnaShaleva) [inline pkg/interop/blockchain/blockchain.go:16]: We can also change `blockchain.GetTransaction(hash []byte)`, `blockchain.GetTransactionHeight(hash []byte)` and `blockchain.GetContract(scriptHash []byte)` signatures.

**Review by AnnaShaleva — CHANGES_REQUESTED** [review]

File: pkg/interop/blockchain/blockchain.go (hunk 2 above)
Comment (fyrchik) [reply pkg/interop/blockchain/blockchain.go:16]: Yeah, fixed.

**Review by AnnaShaleva — APPROVED** [review]

**Comment by roman-khimov** [comment]
Can we implicitly generate `InteropInterface` for storage.Context/iterator.Iterator and similar? Like `x(InteropInterface) Boolean` from `func X(storage.Context) bool`.

**Comment by fyrchik** [comment]
@roman-khimov I have thought about this, but we need to enumerate all types from `interop` we want to replace, because some of them (e.g. `Block`, `Transaction` are valid structures). Maybe we can also check if `interop` type is an empty struct (holds for an actual interops).

**Comment by roman-khimov** [comment]
We have something like three of such cases, enumeration is fine I think.

File: pkg/compiler/debug.go [hunk 3]
```diff
@@ -201,6 +201,21 @@ func (c *codegen) scTypeFromExpr(typ ast.Expr) string {
 	if c.typeOf(typ) == nil {
 		return "Any"
 	}
+	if named, ok := t.(*types.Named); ok {
+		if isInteropPath(named.String()) {
+			name := named.Obj().Name()
+			pkg := named.Obj().Pkg().Name()
+			switch pkg {
+			case "blockchain", "contract":
+				return "Array" // Block, Transaction, Contract
```
Comment (roman-khimov) [inline pkg/compiler/debug.go:210]: But they're `Struct` in neo-go, even though syscalls return `Array`.

Comment (fyrchik) [reply pkg/compiler/debug.go:210]: I think, interop methods signature must be altered to return pointers, after it was made possible.

Comment (roman-khimov) [reply pkg/compiler/debug.go:210]: Yeah, that would be ideal, avoiding conversions and matching this interface with other languages (suppose you want to expose some method that accepts a `Transaction`, C# code would pass an `Array` for it while we would be expecting a `Struct`).

**Review by roman-khimov — CHANGES_REQUESTED** [review]
2c585d6ab2dfec5f9c23de806af83661af10fba8 is a part of f71018c4b5964fb5b70cab23f80ccd5450c86fbd and d5ed9bb4a8a2c94e5b09f7a0b8b7bbffe5d0ea94 should remove ConvertResultToStruct logic (and technically precceed associated manifest generation changes.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**